core/common/src/main/java/alluxio/AbstractClient.java

```diff
@@ -259,12 +257,15 @@ public synchronized void connect() throws AlluxioStatusException {
               retryPolicy.getAttemptCount()));
     }
 
-    if (connectFailReason != null) {
-      throw connectFailReason;
+    /**
+     * Throw as-is if {@link UnauthenticatedException} occurred.
+     */
+    if (lastConnectFailure instanceof UnauthenticatedException) {
+      throw (AlluxioStatusException) lastConnectFailure;
     }
 
     throw new UnavailableException(String.format("Failed to connect to %s @ %s after %s attempts",
-        getServiceName(), mAddress, retryPolicy.getAttemptCount()));
+        getServiceName(), mAddress, retryPolicy.getAttemptCount()), lastConnectFailure);
```

Comment:
Is last connect failure guaranteed to be non null?

Reply:
No. However it's allowed to have null in the places it has been used.